### DISCUSSION TITLE
LazyExtras markdown; How to configure markdownlint-cli2?

### DISCUSSION THREAD

**Discussion by jpridemore-allegion** [LazyVim/LazyVim#4094] [Q&A]
I'm purposely naming this very similar to the discussion for [the previous question](https://github.com/LazyVim/LazyVim/discussions/2268), but it looks like [3 weeks ago](https://github.com/LazyVim/LazyVim/pull/3843) a PR was merged that switched from `markdown-cli` to `markdown-cli2` for performance reasons and it broke the answer to the linked discussion on how to configure it.

Whereas [markdown-cli](https://github.com/igorshubovych/markdownlint-cli?tab=readme-ov-file#usage) has quite a few cli options you can send in, it seems [markdown-cli2](https://github.com/igorshubovych/markdownlint-cli?tab=readme-ov-file#usage) is severely limited in functionality from the command line, and allows no rules disabled like that. It does, however, allow a `--config` option which I can get working. Just to demonstrate the difference of sending in the `--config` option or not:

```shell
riobe@~/notes 
$ markdownlint-cli2 todo.md 
markdownlint-cli2 v0.13.0 (markdownlint v0.34.0)
Finding: todo.md
Linting: 1 file(s)
Summary: 2 error(s)
todo.md:1 MD041/first-line-heading/first-line-h1 First line in a file should be a top-level heading [Context: "Not a header"]
todo.md:7:81 MD013/line-length Line length [Expected: 80; Actual: 184]
riobe@~/notes 
$ markdownlint-cli2 --config ~/.markdownlint-cli2.yaml todo.md 
markdownlint-cli2 v0.13.0 (markdownlint v0.34.0)
Finding: todo.md
Linting: 1 file(s)
Summary: 1 error(s)
todo.md:1 MD041/first-line-heading/first-line-h1 First line in a file should be a top-level heading [Context: "Not a header"]
```

If I include into my config file `lua/plugins/nvim-lint.lua` the following configuration (based off of the answer to the previous discussion):

```lua
return {
  {
    "mfussenegger/nvim-lint",

    opts = {
      linters = {
        -- This does nothing
        markdown = {
          args = { "--config", "~/.markdownlint-cli2.yaml", "--" },
        },
        -- This does nothing
        markdownlint = {
          args = { "--config", "~/.markdownlint-cli2.yaml", "--" },
        },
        -- This stops _all_ linting from running on markdown.
        ["markdownlint-cli2"] = {
          args = { "--config", "~/.markdownlint-cli2.yaml", "--" },
        },
      },
    },
  },
}
```

...I now get no output at all. Only the `markdownlint-cli2` line does anything at all, and I'm assuming it causes an error that gets swallowed somewhere, but I don't really know as there's no output and no diagnostics anymore.

I like the idea of a markdown linter, but I _really_ want to get rid of `MD013` which specifies a line length of 80 character, and I'd _like_ to be able to configure it more generally. I'm not sure why this doesn't work, can anyone offer help?

**Comment by jpridemore-allegion** [comment]
Just to be thorough, I've also tried configuring `markdownlint-cli2` like this too:

```lua
        ["markdownlint-cli2"] = {
          config = { "~/.markdownlint-cli2.yaml" },
        },
```

And like this:

```lua
        ["markdownlint-cli2"] = {
          config = { MD013 = false },
        },
```

These configurations don't break it, but they also don't do anything.

**Comment by luigir-it** [comment]
The correct syntax is this:
```lua
return {
  "mfussenegger/nvim-lint",
  optional = true,
  opts = {
    linters = {
      ["markdownlint-cli2"] = {
        args = { "--config", "/home/pcino/.markdownlint-cli2.yaml", "--" },
      },
    },
  },
}
```

The issue is that "~" is not expanded. And this is the standard behavior in bash when quoted. I also tried with $HOME, which is also not expanded. This hints that the string is passed as literal to the shell. If you want your config to be machine agnostic, you can declare a local variable and substitute it in the args table.
```lua
local HOME = os.getenv("HOME")
return {
  "mfussenegger/nvim-lint",
  optional = true,
  opts = {
    linters = {
      ["markdownlint-cli2"] = {
        args = { "--config", HOME .. "/.markdownlint-cli2.yaml", "--" },
      },
    },
  },
}
```

The $HOME/.markdownlint-cli2.yaml looks like this:
```yaml {filename=".markdownlint-cli2.yaml"}
config:
  MD013: false
```

___ 

- If you want to customize rules for a specific directory, you can create `.markdownlint-cli2.yaml` inside that directory
- If you want to customize rules for a specific file, you can add headings like 
  - \ or
  - \

**Reply by bcvanmeurs** [reply]
Thanks, I got this to work, and you can actually use `vim.fn.expand` to use `~`.
`vim.fn.expand("~/.config/nvim/lua/plugins/.markdownlint-cli2.yaml")`.

One thing I don't really understand is, should I configure this with conform, none-ls or nvim-lint?
I tried using none-ls:

```yaml
local lint = null_ls.builtins.diagnostics
local null_ls_sources = {
  lint.markdownlint_cli2.with({
    extra_args = { "--config", vim.fn.expand("~/.config/nvim/lua/plugins/.markdownlint-cli2.yaml") },
  }),
}
return {
  {
    "nvimtools/none-ls.nvim",
    opts = {
      sources = null_ls_sources,
    },
}
```
But this doesn't seem to work...
(I have `{ import = "lazyvim.plugins.extras.lang.markdown" },` in the lazy setup.)

I am trying to understand what to use when....

**Reply by richtong** [reply]
yes I'm completely confused what is actaully calling markdownlint

**Reply by atkr** [reply]
if you are using LazyVim and have the markdown LazyExtra enabled, the linting is done with nvim-lint. You can see it in the actual markdown.lua file that is being imported as the extra:
```lua
  {
    "mfussenegger/nvim-lint",
    optional = true,
    opts = {
      linters_by_ft = {
        markdown = { "markdownlint-cli2" },
      },
    },
  },
```

**Reply by svenXY** [reply]
hi @atkr , thanks for pointing out the correct syntax for LazyVim - question that remains here is then: how would I add the args that have been suggested above in this case?

**Reply by atkr** [reply]
As posted in the parent comment of this thread @svenXY -> [here](https://github.com/LazyVim/LazyVim/discussions/4094#discussioncomment-10178217)

This is the relevant part of my config, hope it helps.
```lua
return {
  'mfussenegger/nvim-lint',
  opts = {
    linters_by_ft = {
      markdown = { "markdownlint-cli2" },
    },
    linters = {
      ['markdownlint-cli2'] = {
        args = {
          '--config',
          vim.fn.stdpath('config') .. '/lua/plugins/cfg_linters/global.markdownlint-cli2.jsonc',
          '--',
        },
      },
    },
  },
}
```

**Reply by svenXY** [reply]
all right, thanks - I had some problems bringing the two together, with the "extra-challenge" that hyphens are not allowed in keys. It works now, cheers

**Reply by nickjj** [reply]
Thanks a lot, this works great.

Btw @luigir-it you have a typo in this line `<!-- markdownlint-configure-file { "MD013": { "line_lenght": 100} } -->` , it should be `line_length` instead of `line_lenght`.

**Reply by dpetka2001** [reply]
From taking a look through the [docs](https://github.com/DavidAnson/markdownlint-cli2) and also this [maintainer's comment](https://github.com/DavidAnson/markdownlint-cli2/issues/7#issuecomment-695812069) I don't believe so. The other alternative would be to use comment directives to disable rules for a specific file. 

Or you could use the `extends` property at the configuration file of a project, so that your custom settings will be extended by the project's settings. First your custom configuration file will take effect and then the project's configuration file will get merged with yours and subsequent values in project's configuration file will override your custom configuration file. Example [here](https://github.com/DavidAnson/markdownlint/blob/v0.32.1/schema/.markdownlint.jsonc).

Never actually tried this, so take everything I've said with a grain of salt and test it out yourself to make sure you get the behavior you want.

**Reply by wohltat** [reply]
> The $HOME/.markdownlint-cli2.yaml looks like this:
> 
> ```yaml
> config:
>   MD013: false
> ```

Just wanted to write that adding the .yaml file was enough for me to disable the linter warnings. I didn't create any extra .lua config.

**Comment by awilkins** [comment]
Returning to this one because ... you fix the indents ... then the `conform.nvim` prettifier unfixes them and then you
have a lot of warnings about indents again because they're all wrong :-)

What I kinda really want to happpen here is for it to use the `.editorconfig` indent setting for Markdown, but baby steps..

Happily the prettifier also uses `markdownlint-cli2`.

```yaml
# ~/.markdownlint-cli2.yaml
config:
  default: true
  MD007:
    indent: 4
```

```lua
-- ~/.config/nvim/lua/plugins/markdown.lua
return {
  {
    "mfussenegger/nvim-lint",
    opts = {
      linters = {
        ["markdownlint-cli2"] = {
          args = { "--config", vim.fn.expand("$HOME/.markdownlint-cli2.yaml"), "--" },
        },
      },
    },
  },
  {
    "stevearc/conform.nvim",
    opts = {
      formatters = {
        ["markdownlint-cli2"] = {
          args = { "--config", vim.fn.expand("$HOME/.markdownlint-cli2.yaml"), "--fix", "$FILENAME" },
        },
      },
    },
  },
}
```

**Comment by JGillette71** [comment]
Can't figure out why this doesn't work for me...

I want a global solution to disable MD013. I searched through docs, read this and many other discussions, and eventually stumbled across [Follow markdown standards with markdowlint markdownlint-cli2 and nvim-lint](https://www.youtube.com/watch?v=LsABH1_Ma04) on Youtube and @linkarzu's GitHub that contained a solution that is meant to work via the nvim-lint and markdownlint-cli2 plugins. Following these instructions I added `~/Work/dotfiles/.markdownlint.yaml` and pointed to it with `~/.config/nvim/lua/plugins/nvim-lint.lua`, a snippet of which is detailed below.

```lua
local cfg_path = os.getenv("HOME") .. "/Work/dotfiles/.markdownlint.yaml"

return {
  "mfussenegger/nvim-lint",
  optional = true,
  cond = function()
    if vim.fn.filereadable(cfg_path) == 1 then
      return true
    else
      vim.schedule(function()
        vim.notify(
          string.format(
            "[nvim-lint] Skipping markdownlint-cli2 setup — expected config file not found at:\n%s",
            cfg_path
          ),
          vim.log.levels.WARN
        )
      end)
      return false
    end
  end,
  opts = {
    linters = {
      -- https://github.com/LazyVim/LazyVim/discussions/4094#discussioncomment-10178217
      ["markdownlint-cli2"] = {
        args = { "--config", cfg_path, "--" },
      },
    },
  },
}
```

Despite this effort I'm still getting `error MD013/line-length Line length [Expected: 80; Actual: 159] markdownlint` in my markdown files...

Moreover, I'm still seeing the following snippet in `:ConformInfo`. I tried creating `~/.markdownlint-cli2.yaml` and that didn't work either. 

```plaintext
Log file: /home/jbuhg/.local/state/nvim/conform.log
              at async file:///home/jbuhg/.local/share/nvim/mason/packages/markdownlint-cli2/node_modules/markdownlint-cli2/markdownlint-cli2-bin.mjs:14:22 {
            [cause]: Error: ENOENT: no such file or directory, open '/home/jbuhg/.markdownlint-cli2.yaml'
                at async open (node:internal/fs/promises:641:25)
                at async Object.readFile (node:internal/fs/promises:1279:14)
                at async readOptionsOrConfig (file:///home/jbuhg/.local/share/nvim/mason/packages/markdownlint-cli2/node_modules/markdownlint-cli2/markdownlint-cli2.mjs:155:27)
                at async main (file:///home/jbuhg/.local/share/nvim/mason/packages/markdownlint-cli2/node_modules/markdownlint-cli2/markdownlint-cli2.mjs:973:9)
                at async file:///home/jbuhg/.local/share/nvim/mason/packages/markdownlint-cli2/node_modules/markdownlint-cli2/markdownlint-cli2-bin.mjs:14:22 {
              errno: -2,
              code: 'ENOENT',
              syscall: 'open',
              path: '/home/jbuhg/.markdownlint-cli2.yaml'
            }
          }
```

Is this from a redundant lint? Am I writing configs to the wrong locations? Am I overlooking a syntax issue? This has me stumped. I'm a recent convert from VSCode, and unfortunately issues like this have crushed my productivity. Any direction on this matter would be greatly appreciated.

**Reply by merikan** [reply]
@JGillette71 I don't know why your code isn't working but I have a global `.markdownlint.yaml` file and this is my config
```lua
local XDG_CONFIG_DIR = os.getenv("XDG_CONFIG_HOME") or os.getenv("HOME") .. "/.config"

return {
  {
    "mfussenegger/nvim-lint",
    optional = true,
    opts = {
      linters = {
        ["markdownlint-cli2"] = {
          args = {
            "--config",
            XDG_CONFIG_DIR .. "/markdownlint/.markdownlint.yaml",
            "--",
          },
        },
      },
    },
  },
}
```

and my `.markdownlint.yaml` file
```yaml
# By default, enable all rules provided by the linter
default: true

# MD013: Line length
MD013:
  # Set the maximum allowed line length
  line_length: 120
  # Disable the check for specific elements that frequently exceed the limit
  code_blocks: false
  tables: false
  headings: false
```

**Comment by d4rkd3v1l** [comment]
I'm about 100% certain this was working just recently, but now it doesnt anymore... were there any breaking changes recently 0_o


```lua
return {
  "mfussenegger/nvim-lint",
  optional = true,
  opts = {
    linters = {
      ["markdownlint-cli2"] = {
        args = { "--config", os.getenv("HOME") .. "/.config/nvim/.markdownlint.yaml", "--" },
      },
    },
  },
}
```

**Reply by evokateur** [reply]
@d4rkd3v1l I have *just* moidered my `MD013` warnings (for now..?) and my config is very similar, except I use `vim.fn.stdpath("config")` to start at `~/.config/nvim`:

```lua
return {
  "mfussenegger/nvim-lint",
  optional = true,
  opts = {
    linters = {
      ["markdownlint-cli2"] = {
        args = {
          "--config",
          vim.fn.stdpath("config") .. "/.markdownlint-cli2.yaml",
          "--",
        },
      },
    },
  },
}
```

**Reply by evokateur** [reply]
However, I tried your config (verbatim) and it worked, so it must be something else..

```diff
diff --git a/lua/plugins/nvim-lint.lua b/lua/plugins/nvim-lint.lua
index 011588b..0e16fdc 100644
--- a/lua/plugins/nvim-lint.lua
+++ b/lua/plugins/nvim-lint.lua
@@ -4,11 +4,7 @@ return {
   opts = {
     linters = {
       ["markdownlint-cli2"] = {
-        args = {
-          "--config",
-          vim.fn.stdpath("config") .. "/.markdownlint-cli2.yaml",
-          "--",
-        },
+        args = { "--config", os.getenv("HOME") .. "/.config/nvim/.markdownlint.yaml", "--" },
       },
     },
   },
```

**Reply by d4rkd3v1l** [reply]
I'm running out of ideas here...

Running `markdownlint-cli2` with `--fix` param from nvim command mode works:
`:!markdownlint-cli2 --config ~/.config/nvim/.markdownlint.yml --fix %`

Running it via nvim-lint works without `args`, however as soon as I add (even empty) `args`, it stops working 🤷‍♂️
```lua
return {
  "mfussenegger/nvim-lint",
  optional = true,
  opts = {
    linters = {
      ["markdownlint-cli2"] = {
        -- args = { "--config", os.getenv("HOME") .. "/.config/nvim/.markdownlint.yaml", "--" },
        -- args = {},
      },
    },
  },
}
```

I know, that --fix is more like, what nvim-conform will do, however, even without, it should at least show the issues...

**Comment by defields923** [comment]
I have literally tried everything in this thread and about 10 other solutions, I'm pulling my hair out at this point. Why is it so hard to disable a single markdownlint rule? XD

**Comment by dpetka2001** [comment]
Hey everyone. There was a recent change in `nvim-lint` for [markdownlint-cli2](https://github.com/mfussenegger/nvim-lint/commit/ca6ea12daf0a4d92dc24c5c9ae22a1f0418ade37). 

It seems the config has to be changed into following
```lua
return {
  {
    "mfussenegger/nvim-lint",
    opts = {
      linters = {
        ["markdownlint-cli2"] = {
          -- args = { "-", "--config", vim.fn.stdpath("config") .. "/.markdownlint-cli2.yaml", "--" },
          prepend_args = { "--config", vim.fn.stdpath("config") .. "/.markdownlint-cli2.yaml", "--" },
        },
      },
    },
  },
}
```
Notice the extra `"-"` as the first argument in `args`.

PS: I changed the config to use `prepend_args` instead of `args` because this way LazyVim extends the default `args` that are defined in `nvim-lint` for each linter (if any) with our custom ones in `prepend_args`. I commented the default answer I gave, but still left it in case someone wants to use that. Someone should be aware that when we define `args` in our personal configuration the default `args` will be overwritten because of how `vim.tbl_deep_extend` works with list-like values. So, one should also read through the configuration of each linter in `nvim-lint` and check if there are any default `args` that should be included when he defines his own `args` in his personal configuration.

PS2: @d4rkd3v1l, @defields923 pinging in reply to the latest comments of yours.

**Reply by d4rkd3v1l** [reply]
Awesome! Thank you so much 🥳
Works like a charm again, now 😊

```lua
...
    linters = {
      ["markdownlint-cli2"] = {
        prepend_args = { "--config", os.getenv("HOME") .. "/.config/nvim/.markdownlint.yaml", "--" },
      },
    },
...
```

**Reply by richardguinness** [reply]
In case this helps someone else who'd made the dumb mistake I'd made: I had previously placed the nvim-lint block and options in  `$HOME/.config/nvim/lua/config/lint.lua` but it should have been in `$HOME/.config/nvim/lua/plugins/lint.lua`. Relocating the file resolved the issue for me (having made the change above too).

**Comment by awilkins** [comment]
Arrived here :

Literally just put a `.markdownlint.yaml` in my home folder and it worked, no config (I grepped my config folder for it to make sure I'd not already configured it ..)

**Reply by yakimant** [reply]
Hm, I think it should not work:
https://github.com/DavidAnson/markdownlint-cli2/issues/651

Whats your `markdownlint-cli2 --version`?

**Reply by awilkins** [reply]
Mason package file says markdownlint-cli2 version is 0.21.0

Confirmed this worked again on another machine by moving the existing `~/.markdownlint-cli2.yaml` file to `~/.markdownlint.yaml` and now the indent level is respected.

**Reply by DocAmaroo** [reply]
I confirm it works, i'm using v0.22.1

It also have the behavior to overrides custom configuration like presented earlier in this thread
```lua
linters = {
  ["markdownlint-cli2"] = {
    prepend_args = { "--config", vim.fn.stdpath("config") .. "/.markdownlint.yaml", "--" },
  },
},
```
